Adds single, non-generated K8s module

chouseknecht opened the pull request:
##### SUMMARY
Adds new modules `k8s.py`

- Uses the [OpenShift Python client](https://github.com/openshift/openshift-restclient-python)
to provides a single module for managing Kubernetes objects
- Not generated!
- Puts all the Ansible code in one place!
- Fully documented!
- Deprecates [ansible-kubernetes-modules](https://github.com/ansible/ansible-kubernetes-modules). We really don't need hundreds of generated modules that all do exactly the same thing.
- Makes me happy!

Requires #33646

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
k8s.py

##### ANSIBLE VERSION
```
2.5.0
```

Commit:
Adds k8s.py

fabianvf commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,529 @@
+        try:
+            helper = KubernetesAnsibleModuleHelper(api_version, kind)
+            helper.get_model(api_version, kind)
+        except:
```
Bare excepts scare me. Also, there are resource names which appear in both openshift and kubernetes, and I think we need to provide a way for the user to determine which project's API they want to use. This approach could be a sane default though.

fabianvf commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,529 @@
+    name: web
+'''
+
+RETURN = '''
```
Just noting that this RETURN is incomplete

fabianvf commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,529 @@
+    state: absent
+    api_version: v1
+    kind: service
+    namesapce: testing
```
namespace

chouseknecht replied:
Fixed

fabianvf commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,529 @@
+            try:
+                helper = OpenShiftAnsibleModuleHelper(api_version, kind)
+                helper.get_model(api_version, kind)
+            except:
```
Just noting the second bare except

fabianvf commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,529 @@
+      apiVersion: v1
+      kind: Service
+      metadata:
+        name: web
```
Curious what the behavior would be if the name/namespace/apiversion/kind were provided at the top level and here and did not match.

chouseknecht replied:
Great question. The individual values of `kind`, `name`, `namespace`, `api_version` will be overwritten by those found in a `resource_definition` or a config read in from `src`.

I updated the doc strings for both the `resource_definition` and `src` options. 

Thanks!

fabianvf commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,529 @@
+                raise
+        return helper
+
+    def execute_module(self):
```
Do you think any of this logic belongs in the AnsibleModuleHelper mixin?

chouseknecht replied:
Maybe. I'm open to a discussion. I copied 90% of this module from `module_utils/k8s_common.py` and `module_utils/openshift_common.py`.

chouseknecht replied:
@fabianvf 

I think the more code we put on the Ansible side the better, simply because Ansible has such a large community. We're more likely to increase adoption and collaboration, if a reasonable amount of code lives in the Ansible world.

fabianvf reviewed (commented):
Like the approach, left a couple questions

chouseknecht commented:
Here's my testing playbook so far, that runs against a local minishift instance:

```
- name: Test ansible-k8s-module
  hosts: localhost
  connection: local

  tasks:
  - name: Create a project
    k8s:
      name: testing
      kind: project
      api_version: v1
    register: output

  - name: show output
    debug:
      var: output

  - name: Create a service
    k8s:
      state: present
      resource_definition:
        apiVersion: v1
        kind: Service
        metadata:
          name: web
          namespace: testing
          labels:
            app: galaxy
            service: web
        spec:
          selector:
            app: galaxy
            service: web
          ports:
          - protocol: TCP
            targetPort: 8000
            name: port-8000-tcp
            port: 8000
    register: output

  - name: show output
    debug:
      var: output

  - name: Delete an existing PVC
    k8s:
      name: web
      namespace: testing
      api: v1
      kind: persistent_volume_claim
      state: absent

  - name: Create PVC
    k8s:
      state: present
      inline: &pvc
        apiVersion: v1
        kind: PersistentVolumeClaim
        metadata:
          name: elastic-volume
          namespace: testing
        spec:
          resources:
            requests:
              storage: 5Gi
          accessModes:
          - ReadWriteOnce

  - name: Show output
    debug:
      var: output

  - name: Create PVC
    k8s:
      state: present
      inline: *pvc
    register: output

  - name: Show output
    debug:
      var: output

  - name: PVC should be idempotent
    assert:
      that: not output.changed

  - name: Delete an existing deployment config
    k8s:
      name: elastic
      namespace: testing
      api: v1
      kind: DeploymentConfig
      state: absent

  - name: Create deployment config
    k8s:
      state: present
      inline: &dc
        apiVersion: v1
        kind: DeploymentConfig
        metadata:
          name: elastic
          labels:
            app: galaxy
            service: elastic
          namespace: testing
        spec:
          template:
            metadata:
              labels:
                app: galaxy
                service: elastic
            spec:
              containers:
              - name: elastic
                volumeMounts:
                - mountPath: /usr/share/elasticsearch/data
                  name: elastic-volume
                command: ['elasticsearch']
                image: 'ansible/galaxy-elasticsearch:2.4.6'
              volumes:
              - name: elastic-volume
                persistentVolumeClaim:
                  claimName: elastic-volume
          replicas: 1
          strategy:
            type: Rolling
    register: output

  - name: Show output
    debug:
      var: output

  - name: Create deployment config
    k8s:
      state: present
      inline: *dc
    register: output

  - name: Show output
    debug:
      var: output

  - name: Deployment config should be idempotent
    assert:
```

Commit:
Two module approach to support overlapping objects

chouseknecht commented:
@fabianvf 

Switched to a two module approach to address object name collision between OpenShift and K8s. And to keep things DRY, moved the code back into module_utils.

The new `module_utils/k8s_common.py` and `module_utils/openshift_common.py` are not compatible with the generated modules. The idea of course is to get rid of them, and I'm hoping we can do that quickly.

smarterclayton commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,111 @@
+  k8s:
+    state: absent
+    api_version: v1
+    kind: service
```
If you pipe this through to the URL, then this should really be "resource".  In Kube "kind" is the schema, "resource" is the thing you access via a client.  Kind is always a constant (uppercase), resource is always the lowercase RESTful name.

I would recommend `resource: service` (if you're doing discovery) or `resource: services` (if you're hardcoding).  Mapping `kind` to `resource` is something we explicitly do on the client by looking up a Kind in the discovery doc, so overlapping them can be dangerous.

chouseknecht replied:
@smarterclayton 

The `kind` module parameter is one-for-one with the `kind` option found in a Kubernetes config file. In both cases the value of `kind` is a model name.

Normally, in a config file, the model name is in uppercase. We're allowing, however, for the fact that the user will likely provide a lowercased value, and thus the lowercase `service` or `namespace` value.

I updated the doc strings to reflect exactly what the `kind` parameter is: https://github.com/chouseknecht/ansible/blob/feature/k8s_module/lib/ansible/utils/module_docs_fragments/kubernetes.py#L63

Hopefully that clarifies things, and lines up with your thinking.

smarterclayton commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,111 @@
+  k8s:
+    name: testing
+    api_version: v1
+    kind: namespace
```
Same comment as below.

smarterclayton commented on lib/ansible/utils/module_docs_fragments/kubernetes.py:
```diff
@@ -0,0 +1,116 @@
+      variable.
+  cert_file:
+    description:
+    - Path to a certificate used to authenticate with the API. Can also be specified via K8S_AUTH_CERT_FILE environment
```
These are specific env vars created for ansible?  I assume there's a spec that covers this - we have explicitly not added them to Kube for reasons, but they seem fine to use from ansible.

chouseknecht replied:
@smarterclayton 

The environment vars are introduced by the [Openshift client](https://github.com/openshift/openshift-restclient-python). We're making them available here for convenience.

smarterclayton commented on lib/ansible/utils/module_docs_fragments/kubernetes.py:
```diff
@@ -0,0 +1,116 @@
+      if set to C(absent). A patch operation results in merging lists and updating
+      dictionaries, with lists being merged into a unique set of values. If a list
+      contains a dictionary with a I(name) or I(type) attribute, a strategic merge
+      is performed, where individual elements with a matching I(name_) or I(type)
```
What is type in this context?  Since Kube has 3 different patch strategies (specified via content type), strategic merge patch may be versioned in the future and clients may have to customize it, it might actually be safer to not depend on strategic merge patch at all and instead make the content opaque - something like raw JSON and a content type to pass to the server.

Also, in the future Kube is considering having a "typed Patch" which *might* look something like:

```
kind: Patch
apiVersion: meta.k8s.io/v1
patchType: <mimeType>
patchData: <raw JSON>
target: <reference to an object, including Kind / Name / Namespace>
```

However, that could be further off.  It mirrors the spirit of what I'm inferring from the doc, but could compete / conflict with a smarter Ansible patch.

I do agree in the short term ansible does need some structured patch on this model.

smarterclayton commented:
A few possible places where you may get burned in the future.  The kind / resource distinction is pretty important re: discovery - I think surfacing the kube concept directly is probably the correct thing so that the concepts don't get mushed up.

Commit:
Clarify docs

fabianvf commented:
@chouseknecht  Would we want to support getting/listing resources from openshift?

Commit:
Clary state and force parameters

Commit:
Updates examples

Commit:
Add description and display_name

Commit:
Fix linting issues

chouseknecht commented:
@fabianvf 

You can do the following to get a specific deployment config:

```
- name: Get a deployment config
  openshift:
    api_version: v1
    kind: DeploymentConfig
    name: elastic
    namespace: testing
```

I added back support in the common code for `List` models, and now you can also do this:

```
- name: List deployment configs
      openshift:
        api_version: v1
        kind: DeploymentConfigList
        namespace: testing
      register: output
```

I suppose to the point @smarterclayton was making, our `kind` in the above example is `DeploymentConfigList`, which represents a model or schema. We are being consistent in how we're using `kind`, I think. However, the above example is not great from a usability standpoint, as most users likely don't know model names, and we don't have them documented anywhere.

If this were the `oc` client, for example, the command would be `oc get dc` to get a list of deployment configs. It looks a bit cryptic, but it's well documented, and well known. Ours is... neither.

I think what we have in this iteration is a reasonable way of performing CRUD operations on schema objects. But the syntax may not be intuitive for retrieving lists of objects, nor for performing actions.

Our choice are... write a bunch of documentation that perhaps includes a generated list of schema names? Maybe providing a reference is enough.  Or, another option might be writing a module for handling lists, and yet another module for handling actions.

flaper87 commented:
I like this idea. As someone that has been consuming the generated `ansible-kubernetes` modules, I think this module will bring quite a few benefits. One thing that didn't feel right when I started reviewing this patch is that it is not very different from just writing the `Kubernetes` templates but, tbh, neither are the generated modules. 

I've dropped a comment on #33646 with regards to the use of openshift-restclient vs kubernetes client.

Thanks for working on this, y'all. I'm happy to help maintaining these modules (k8s_common, openshift and k8s) if you'd like to have another pair of hands on it.

dgoodwin commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,126 @@
+    state: present
+    definition:
+      apiVersion: v1
+      kind: Service
```
Really like this raw access to the kubernetes yaml, personal preference I don't want a layer of declarative config on top of a layer of declarative config so it's great to see this available.

dgoodwin commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,126 @@
+
+
+def main():
+    KubernetesAnsibleModule().execute_module()
```
Am I correct that we've still got no tie-in with the advanced 3-way merge logic currently baked into the kubectl/oc client binaries? That functionality seems so important for Ansible, and rumored to head for the API server soon but still may be a few releases out.

chouseknecht replied:
@dgoodwin 

It doesn't do a 3-way merge. It merges an existing object with the requested params, then compares the merged object to the latest object. If there are differences, it makes makes a request to the API (patch, update, post, etc.), and returns 'changed', along with the new object. If no differences, then  no request is sent to the API, and OK is returned, along with the current object.

The modules call the API, rather than relying directly on `oc` or `kubectl`. So when the 3-way logic does land on the server, we'll want to switch to letting the API handle merging and comparing objects.

dgoodwin commented on lib/ansible/modules/clustering/k8s.py:
```diff
@@ -0,0 +1,126 @@
+- name: Create a Service object by reading the definition from a file
+  k8s:
+    state: present
+    src: /testing/service.yml
```
Also like having the option of using a file on control host, or inline with "definition". Is src/definition a pattern used like that anywhere else in ansible or are they named otherwise?

dgoodwin commented on lib/ansible/utils/module_docs_fragments/kubernetes.py:
```diff
@@ -0,0 +1,128 @@
+    - Path to an existing Kubernetes config file. If not provided, and no other connection
+      options are provided, the openshift client will attempt to load the default
+      configuration file from I(~/.kube/config.json). Can also be specified via K8S_AUTH_KUBECONFIG environment
+      variable.
```
Would it be useful to allow inline kubeconfig definition as well similar to the src/definition pattern?

chouseknecht replied:
Not sure what you mean. There is both a `resource_definition` parameter, and a `src` parameter. The `resource_definition` can be used to pass an object definition (e.g. the Yaml describing an object, such as a Service, Resource, DeploymentConfig, etc.), and the `src` parameter can be used to read the object definition from a file.

dgoodwin replied:
Referring specifically to the kubeconfig definition, the remote file path will cover a lot of use cases but I was wondering if it would be useful to be able to specify it inline as well, which would open the door for keeping your authentication with your ansible playbooks/roles.

dgoodwin reviewed (commented):
If interested there may be some additional ideas available here in a module we were playing with: https://github.com/dgoodwin/kubectl-ansible/blob/master/README.md

There's a little code but mostly just a picture of what it could look like. This approach leverages the kubectl binary as the API, so we can get at all that juicy apply logic that seems really, really important for Ansible. (until such a time as it's available in the Kube API server)

Few other concepts listed in there for making subtle edits to your kubeconfig to apply without requiring managing your own temporary templated files.

maxamillion commented:
shipit

jimi-c reviewed (commented):
:+1:
